docs/faq.md

```diff
@@ -277,3 +284,25 @@ module.exports = {
   ],
 };
 ```
+
+## How to use TypeORM in ESM projects?
+
+Make sure to add `"type": "module"` in the `package.json` of your project so TypeORM will know to use `import( ... )` on files.
+
+To avoid circular dependency import issues use the `Related` wrapper type for relation type definitions in entities:
+
+```typescript
+@Entity()
+export class User {
+
+    @OneToOne(() => Profile, profile => profile.user)
+    profile: Related<Profile>;
+
+}
+```
+
+Doing this prevents the type of the key from being saved in the transpiled code in the key metadata, preventing circular dependency imports.
```

Comment:
can you please explain me what exact problem with the "key" stored in the transpiled code and what is called key? Is it "profile"?

Reply:
Let's say we have these TS ESM files:
```typescript
// Profile.ts

import {Entity, OneToOne} from "typeorm";
import User from "./User.js";

@Entity()
export default class Profile {
    @OneToOne(() => User, user => user.profile)
    user: User;
}
```
```typescript
// User.ts

import {Entity, OneToOne} from "typeorm";
import Profile from "./Profile.js";

@Entity()
export default class User {
    @OneToOne(() => Profile, profile => profile.user)
    profile: Profile;
}
```

Their compiled code will look like this:
<details>
    <summary> Profile.js </summary>

```javascript
// Profile.js

var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Entity, OneToOne } from "typeorm";
import User from "./User.js";
let Profile = class Profile {
};
__decorate([
    OneToOne(() => User, user => user.profile),
    __metadata("design:type", User)
], Profile.prototype, "user", void 0);
Profile = __decorate([
    Entity()
], Profile);
export default Profile;
//# sourceMappingURL=Profile.js.map
```
</details>

<details>
    <summary> User.js </summary>

```javascript
// User.js

var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Entity, OneToOne } from "typeorm";
import Profile from "./Profile.js";
let User = class User {
};
__decorate([
    OneToOne(() => Profile, profile => profile.user),
    __metadata("design:type", Profile)
], User.prototype, "profile", void 0);
User = __decorate([
    Entity()
], User);
export default User;
//# sourceMappingURL=User.js.map
```
</details>

And when we run the code we will get this error:
```
file:///Users/user/Documents/workspace/project/dist/entity/User.js:17
    __metadata("design:type", Profile)
                              ^

ReferenceError: Cannot access 'Profile' before initialization
    at file:///Users/user/Documents/workspace/project/dist/entity/User.js:17:31
    at ModuleJob.run (node:internal/modules/esm/module_job:195:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:331:24)
    at async Command.<anonymous> (file:///Users/user/Documents/workspace/project/dist/index.js:29:34)

```

The issue is that the statement `__metadata("design:type", Profile)` on `User.js` access `Profile`
while the statement `__metadata("design:type", User)` on `Profile.js` access `User`, thus creating a circular dependency between 2 files that require the other file to be already loaded in order to finish loading itself.

Since in these types of cases the metadata of the property type is redundant since the relation decorator already requires you to provide the type of the column (via `() => Profile` in this example), removing the property type information will not damage any functionality.

This is [a known issue](https://www.npmjs.com/package/reflect-metadata#:~:text=have%20issues%20with-,circular,-or%20forward%20references) with `tsconfig.json`'s `emitDecoratorMetadata`, fixing it on the TypeScript level is very complicated so I gave up on that.